Fix submission dumping with repeating groups

benhowes opened the pull request:
### What is the context of this PR?

1. A minor fix to dumping the submission, because an unset global was relied on. Adds a test for this case.
2. As part of 1, reduces (but doesn't entirely remove) use of `g.schema` throughout. This means passing the schema as an argument a lot more, particularly in the templating where this wasn't done before. This appears to be because until repeating groups were added, there was no need for knowledge of the schema in these places. 
3. Due to the reduced use of `g.schema`, `_answer_is_in_repeating_group` has been removed from `rules.py`, which has required modification of a boat load of tests since this was mocked in many places. In general, I've tried to patch a higher level call - i.e. mock out `evaluate_goto` instead of the low level one. This is to try to isolate units more.
4. A couple of tests have been removed, either were they were testing another unit by proxy, and in one case there was an identical test under two names.
5. A couple of tests have been added - these are unrelated to the fix, however this PR was failing on code coverage, due to removing tested code (which increased the percentage of lines which were not tested).

This PR has ended up being bigger than intended. Some of the unit tests in templating have changed more than they strictly had to, however since I needed to change all the call signatures anyway, I wanted to isolate the units more whist I was understanding what each test did.

### How to review 
1. Tests should all pass!
2. Running `census_household.json` or other surveys with repeating groups should allow dumping of the submission on `/dump/submission`. To ensure the fix works, test one is failing on master first (should return a 500 error)

benhowes committed:
Fix dumping of submission with repeating groups

Reduces usage of g.schema by passing the schema more.

benhowes commented on tests/integration/mci/test_backwards_navigation_after_submission.py:
```diff
@@ -1,11 +1,13 @@
 
-class TestBackwardsNavigationAfterSubmission(TestHappyPath):
+
+class TestBackwardsNavigationAfterSubmission(IntegrationTestCase, HappyPathHelperMixin):
```
I just wanted to signpost here that the new mixin class is there because previously, when this inherited from `TestHappyPath`, it meant that all the test cases from `TestHappyPath` were also executed here.

jonnyshaw89 replied:
Is this similar to this? https://github.com/ONSdigital/eq-survey-runner/blob/master/tests/app/app_context_test_case.py#L39

benhowes replied:
Yes, except that one must be executed a lot more than the one I've fixed

benhowes replied:
If it's okay with you, I'll put through a separate PR which addresses that. Of course it's being collected because pytest is looking for everything starting with `test_`, so I'll just rename that

insacuri commented on app/templating/summary/question.py:
```diff
@@ -6,14 +6,14 @@
 class Question:
 
-    def __init__(self, question_schema, answer_store, metadata):
+    def __init__(self, question_schema, answer_store, metadata, schema):
```
What's the difference between question_schema and schema?

benhowes replied:
`question_schema` is a 'fragment' (n.b. contentious term) of the whole `schema`, but there are places that the whole schema is needed to evaluate if something's in a repeating group

insacuri replied:
thats what I thought, just wanted to be sure I got it

insacuri commented on app/views/questionnaire.py:
```diff
@@ -258,7 +263,7 @@ def get_thank_you(eq_id, form_type):  # pylint: disable=unused-argument
 
     session_data = get_session_store().session_data
-    g.schema = load_schema_from_session_data(session_data)
+    g.schema = schema = load_schema_from_session_data(session_data)
```
Pretty cool, had no idea that you could do chain assignments in python!

nealedj replied:
Really useful. Don't make the mistake of assigning the same reference to multiple variables and thinking they're different objects.

Pretty obvious:
```
>>> a_string = b_string = "foo"
>>> a_string = "bar"
>>> a_string
'bar'
>>> b_string
'foo'
```

Sometimes less expected:
```
>>> a_list = b_list = []
>>> a_list.append('foo')
>>> a_list
['foo']
>>> b_list
['foo']
```

insacuri reviewed (approved):
LGTM 
EDIT: Though 2 others should look at this cause there's a chance I missed something cause I'm still new to the codebase ^^

benhowes commented:
I missed a few instances of `_answer_is_in_repeating_group` which appeared after rebasing, working through those now

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -125,6 +125,8 @@ def save_questionnaire_store_wrapper(*args, **kwargs):
 def get_block(routing_path, eq_id, form_type, collection_id, group_id, group_instance, block_id):  # pylint: disable=unused-argument,too-many-locals
     current_location = Location(group_id, group_instance, block_id)
+    metadata = get_metadata(current_user)
+    schema = load_schema_from_metadata(metadata)
```
worth considering this as a wrapper? like we do with `@full_routing_path_required`?
given we repeat these lines in most of the views.

benhowes replied:
But I've not included it here because this was already getting pretty big!

jonnyshaw89 replied:
I wasn't thinking specifically about pytest. i was looking more at the repeated lines in every view

benhowes replied:
Shall I put that on the backlog?

jonnyshaw89 replied:
Id say it was a change related to the code changed here?

benhowes replied:
Yeah, okay fair. I'll do that.

jonnyshaw89 reviewed (changes requested):
A nice bit of boy scouting done here 👍

jonnyshaw89 commented:
@benhowes can you try and hold off with squashing until you have 2 ticks? it just makes re-reviews a lot easier as we can just re-review the changes and not have to look over all the files again.

benhowes commented:
I hadn't realised that implication of squashing - sorry!

benhowes committed:
refactored duplicated code into wrappers

jonnyshaw89 commented on app/views/questionnaire.py:
```diff
@@ -123,155 +129,165 @@ def save_questionnaire_store_wrapper(*args, **kwargs):
 @login_required
+@with_answer_store
+@with_metadata
+@with_schema
```
Is the order important here?
Just a thought, can we make the wrappers pass in the value as a kwarg?

benhowes replied:
Order is very important. can easily pass as kwargs, but I was being consistent with the existing `@full_routing_path_required` decorator

jonnyshaw89 replied:
Yeh, just thinking the more we add the more brittle it could become to order, maybe if we make them all use kwarg it would be more resilient?

benhowes replied:
Actually, thinking about it, there would be some interesting errors if they were kwargs and one was missing for some reason, because you can't easily provide valid defaults for those arguments, meaning we'd have to introduce some `if schema is None: ...` for each of the decorators. Unless I'm missing something of course.

jonnyshaw89 replied:
would we also get funky errors if they were in the wrong order? so a schema was actually metadata?

benhowes replied:
Yep, we would. it's weird errors either way, both ways have the same downsides I think. What do you reckon @jonnyshaw89

benhowes committed:
increased coverage

benhowes committed:
linting

benhowes commented:
@kierenbeckett @jonnyshaw89 since you reviewed, I added some tests and removed some unreachable code in this commit: https://github.com/ONSdigital/eq-survey-runner/pull/1694/commits/80920a80caf634bab0c81e98f366b22a02d2b7b9, can you have a look over to ensure you're happy with the changes?

jonnyshaw89 commented:
👍 Happy

kierenbeckett commented:
👍
